File: src/py_pubsub/py_pubsub/topics_practice.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose

class MinimalPublisher(Node):

    def __init__(self):
        super().__init__('publisher')
        self.publisher_ = self.create_publisher(Twist, '/turtle1/cmd_vel', 10)
        self.subscription = self.create_subscription(
            Pose,
            '/turtle1/pose',
            self.pose_callback,
            10)
        self.timer_period = 1.0  # seconds
        self.timer_ = self.create_timer(self.timer_period, self.timer_callback)
        self.count_ = 5  # Countdown starting from 5
        self.drawing_spiral = False
        self.spiral_angle = 30.0  # Angle must be a float
        self.spiral_length = 1.0  # Length must be a float (MODIFICADO)
        self.current_pose = None

    def timer_callback(self):
        if self.count_ > 0:
            self.get_logger().info('Countdown: %d' % self.count_)
            self.count_ -= 1
        elif self.count_ == 0 and not self.drawing_spiral:
            self.get_logger().info('Drawing spiral')
            self.drawing_spiral = True
            self.timer_.cancel()  # Stop the countdown timer
            self.timer_ = self.create_timer(0.1, self.draw_spiral)  # Start drawing spiral

    def pose_callback(self, msg):
        self.current_pose = msg

    def draw_spiral(self):
        if self.current_pose is None:
            return

        twist = Twist()
        twist.linear.x = self.spiral_length
        twist.angular.z = self.spiral_angle
        self.publisher_.publish(twist)
        self.spiral_length += 10.1  # Increase the length for the spiral effect (MODIFICADO)

        # Check boundaries
        if (self.current_pose.x < 1.0 or self.current_pose.x > 10.0 or
                self.current_pose.y < 1.0 or self.current_pose.y > 10.0):
            self.get_logger().info('Going straight')
            self.drawing_spiral = False
            self.timer_.cancel()  # Stop the spiral timer
            self.timer_ = self.create_timer(0.1, self.go_straight)  # Start going straight

    def go_straight(self):
        twist = Twist()
        twist.linear.x = 2.0
        twist.angular.z = 0.0
        self.publisher_.publish(twist)

        if (self.current_pose.x < 1.0 or self.current_pose.x > 10.0 or
                self.current_pose.y < 1.0 or self.current_pose.y > 10.0):
            twist.linear.x = 2.0  # Turn around if hitting the boundary (MODIFICADO)
            self.publisher_.publish(twist)


def main(args=None):
    rclpy.init(args=args)

    minimal_publisher = MinimalPublisher()

    rclpy.spin(minimal_publisher)

    minimal_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/cpp_pubsub/src/topics_practice.cpp
#include <chrono>
#include <memory>
#include <string>
#include <cmath>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "turtlesim/msg/pose.hpp"

using namespace std::chrono_literals;

class MinimalPublisher : public rclcpp::Node
{
public:
    MinimalPublisher()
    : Node("minimal_publisher"), count_(5), spiral_phase_(true), angle_(0.0), radius_(0.1)
    {
        velocity_publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 10);
        pose_subscriber_ = this->create_subscription<turtlesim::msg::Pose>(
            "turtle1/pose", 10, std::bind(&MinimalPublisher::pose_callback, this, std::placeholders::_1));

        timer_ = this->create_wall_timer(
            1s, std::bind(&MinimalPublisher::timer_callback, this));
    }

private:
    void timer_callback()
    {
        if (count_ > 0) {
            RCLCPP_INFO(this->get_logger(), "Countdown: %zu", count_);
            count_--;
        } else {
            if (spiral_phase_) {
                RCLCPP_INFO(this->get_logger(), "Drawing spiral");
                draw_spiral();
            } else {
                RCLCPP_INFO(this->get_logger(), "Going straight");
                go_straight();
            }
        }
    }

    void draw_spiral()
    {
        geometry_msgs::msg::Twist twist;
        twist.linear.x = radius_;
        twist.angular.z = 1.0;
        radius_ += 0.1;         //(MODIFICADO 0.01)
        velocity_publisher_->publish(twist);
    }

    void go_straight()
    {
        geometry_msgs::msg::Twist twist;
        twist.linear.x = 1.0;
        twist.angular.z = 0.0;
        velocity_publisher_->publish(twist);
    }

    void pose_callback(const turtlesim::msg::Pose::SharedPtr msg)
    {
        if (msg->x <= 1.0 || msg->x >= 8.0 || msg->y <= 1.0 || msg->y >= 8.0) {    //(MODIFICADO AREA)
            spiral_phase_ = false;                                                   //if (msg->x <= 1.0 || msg->x >= 10.0 || msg->y <= 1.0 || msg->y >= 10.0) {
        }
    }

    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr velocity_publisher_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_subscriber_;
    size_t count_;
    bool spiral_phase_;
    double angle_;
    double radius_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MinimalPublisher>());
    rclcpp::shutdown();
    return 0;
}
